README.md
# Share content for flutter

[![pub package](https://img.shields.io/pub/v/flutter_share_content.svg)](https://pub.dartlang.org/packages/flutter_share_content)

A simple plugin for Flutter to share content.

This content could be a picture but also a message. Furthermore a title can be added to show in the sharing screen.
 
Every version of Android is supported, included from version 24 and above.
For those devices is mandatory the use of `FileProvider` class in order to grant access 
to the images stored in the phone.

Version 0.2.0 is supporting androidX.

iOS is not going to be implemented because I don't have a macOS.


## Installation

First add `flutter_share_content` as a dependency in your `pubsec.yaml`

### Android

Add this in your `build.gradle`

```gradle
dependencies {
    api 'androidx.legacy:legacy-support-v4:1.0.0'
}
```

And this in your `AndroidManifest.xml`

```xml
<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
<uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>

...

<provider
    android:name="androidx.core.content.FileProvider"
    android:authorities="${applicationId}.provider"
    android:exported="false"
    android:grantUriPermissions="true">
    <meta-data
        android:name="android.support.FILE_PROVIDER_PATHS"
        android:resource="@xml/filepaths"/>
</provider>
```
Also create `res/filepaths.xml` with this content

```xml
<?xml version="1.0" encoding="utf-8"?>
<paths xmlns:android="http://schemas.android.com/apk/res/android">
<external-path name="external_files" path="."/>
</paths>
```


### Example

```dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'dart:io';

import 'package:image_picker/image_picker.dart';
import 'package:flutter_share_content/flutter_share_content.dart';

void main() => runApp(MyApp());


class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: "Share Content example app",
      debugShowCheckedModeBanner: false,
      home: ShareContent(),
    );
  }
}

class ShareContent extends StatefulWidget {
  @override
  _ShareContentState createState() => _ShareContentState();
}

class _ShareContentState extends State<ShareContent> {
  File _image;
  TextEditingController _title = new TextEditingController();
  TextEditingController _msg =  new TextEditingController();
  TextEditingController _subject = new TextEditingController();
  BuildContext _context;

  Future getImage() async {
    File image = await ImagePicker.pickImage(source: ImageSource.gallery);

    setState(() {
      _image = image;
    });
  }

  void getText(){
    showModalBottomSheet(context: _context,
        builder:(BuildContext context) {
          return new Container(
            child: new Column(
              children: <Widget>[
                new TextField(
                  decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    hintText: "Title",
                    contentPadding: EdgeInsets.only(bottom: 50.0, top: 10.0, left: 12.0),
                  ),

                  controller: _title,
                ),
                new TextField(
                  decoration: InputDecoration(
                      border: OutlineInputBorder(),
                      hintText: "Message",
                    contentPadding: EdgeInsets.only(bottom: 150.0, top: 10.0, left: 12.0),
                  ),
                  maxLines: null,
                  keyboardType: TextInputType.multiline,

                  controller: _msg,
                )
              ],
            ),
          );
        }
    );
  }

  void share(){
    FlutterShareContent.shareContent(imageUrl: _image == null? null: _image.path,
        msg: _msg.text, title: _title.text);
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    _context = context;
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Share Content example app'),
        ),
        body: new Center(
          child: _image == null
              ? new Text('No image selected.')
              : new Image.file(_image),
        ),
        floatingActionButton:
        FloatingActionButton(onPressed: share, child: Icon(Icons.share),),

        bottomNavigationBar:
        BottomAppBar(
          child:
          Row(
            children: <Widget>[
              new FloatingActionButton(
                onPressed: getImage,
                tooltip: 'Pick Image',
                child: new Icon(Icons.camera_alt),
              ),
              new FloatingActionButton(
                onPressed: getText,
                tooltip: 'Show text',
                child: new Icon(Icons.text_fields),
              ),
            ],
          )

        ),

      ),
    );
  }
}

```



CHANGELOG.md
## 0.2.1+1

* Updated dependencies

## 0.2.0

* Migrate to AndroidX on android support library. 
If you are using the android support library and you want to use that version, you should migrate to Androidx your project. 

## 0.1.0

* Initial release


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_share_content_example/main.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(ShareContent());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


android/src/main/java/io/flutter/plugins/fluttersharecontent/FlutterShareContentPlugin.java
package io.flutter.plugins.fluttersharecontent;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;

import androidx.core.content.FileProvider;

import java.io.File;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * FlutterShareContentPlugin
 */
public class FlutterShareContentPlugin implements MethodCallHandler {

    private final PluginRegistry.Registrar registrar;

    public FlutterShareContentPlugin(PluginRegistry.Registrar registrar) {
        this.registrar = registrar;
    }

    /**
     * Plugin registration.
     */
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_share_content");
        channel.setMethodCallHandler(new FlutterShareContentPlugin(registrar));
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        if (call.method.equals("getPlatformVersion")) {
            result.success("Android " + android.os.Build.VERSION.RELEASE);
        } else if (call.method.equals("shareContent")) {
            shareContent(call, result);
        } else {
            result.notImplemented();
        }
    }


    private void shareContent(MethodCall call, Result result) {
        String path = call.argument("path");
        String msg = call.argument("msg");
        String title = call.argument("title");
        title = title == null || title.equals("") ? "Share to" : title;

        Intent share = new Intent(Intent.ACTION_SEND);

        if (path != null) {
            imageIntent(share, path);
        }

        if (msg != null) {
            share.putExtra(Intent.EXTRA_TEXT, msg);
        }

        registrar.activity().startActivity(Intent.createChooser(share, title));
        result.success("Sharing...");
    }

    private void imageIntent(Intent share, String path) {
        File media = new File(path);

        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {

            Uri apkURI = FileProvider.getUriForFile(
                    registrar.activity().getBaseContext(),
                    registrar.activity().getApplicationContext()
                            .getPackageName() + ".provider", media);

            share.putExtra(Intent.EXTRA_STREAM, apkURI);
        } else
            share.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(media));

        share.setType("image/*");

        share.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
    }
}


example/lib/main.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_share_content/flutter_share_content.dart';
import 'package:image_picker/image_picker.dart';

void main() => runApp(MyApp());

/// StatelessWidget to show the Stateful one
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: "Share Content example app",
      debugShowCheckedModeBanner: false,
      home: ShareContent(),
    );
  }
}

/// ShareContent to create the state
class ShareContent extends StatefulWidget {
  @override
  _ShareContentState createState() => _ShareContentState();
}

/// Made in in that way to show the modal bottom sheet
class _ShareContentState extends State<ShareContent> {
  File _image;
  TextEditingController _title = new TextEditingController();
  TextEditingController _msg = new TextEditingController();
  BuildContext _context;

  Future getImage() async {
    File image = await ImagePicker.pickImage(source: ImageSource.gallery);

    setState(() {
      _image = image;
    });
  }

  void getText() {
    showModalBottomSheet(
        context: _context,
        builder: (BuildContext context) {
          return new Container(
            child: new Column(
              children: <Widget>[
                new TextField(
                  decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    hintText: "Title",
                    contentPadding:
                        EdgeInsets.only(bottom: 50.0, top: 10.0, left: 12.0),
                  ),
                  controller: _title,
                ),
                new TextField(
                  decoration: InputDecoration(
                    border: OutlineInputBorder(),
                    hintText: "Message",
                    contentPadding:
                        EdgeInsets.only(bottom: 150.0, top: 10.0, left: 12.0),
                  ),
                  maxLines: null,
                  keyboardType: TextInputType.multiline,
                  controller: _msg,
                )
              ],
            ),
          );
        });
  }

  void share() {
    FlutterShareContent.shareContent(
        imageUrl: _image == null ? null : _image.path,
        msg: _msg.text,
        title: _title.text);
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    _context = context;
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Share Content example app'),
        ),
        body: new Center(
          child: _image == null
              ? new Text('No image selected.')
              : new Image.file(_image),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: share,
          child: Icon(Icons.share),
        ),
        bottomNavigationBar: BottomAppBar(
            child: Row(
          children: <Widget>[
            new FloatingActionButton(
              onPressed: getImage,
              tooltip: 'Pick Image',
              child: new Icon(Icons.camera_alt),
            ),
            new FloatingActionButton(
              onPressed: getText,
              tooltip: 'Show text',
              child: new Icon(Icons.text_fields),
            ),
          ],
        )),
      ),
    );
  }
}
